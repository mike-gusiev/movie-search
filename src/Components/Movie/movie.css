.movie {
    flex: 1 0 25%;
    display: flex;
    justify-content: center;
    margin-bottom: 20px;
}

.movie__container:hover .movie__info {
    background: rgba(0, 0, 0, 0.7);
}

.movie__container:hover .movie__title {
    background: none;
}

.movie__container:hover .movie__additional {
    height: 100%;
    background: rgba(0, 0, 0, 0);
}

.movie__container {
    width: 280px;
    height: 400px;
    overflow: hidden;
    position: relative;
}

img {
    width: 100%;
    height: 400px;
    border: 1px solid #ccc;
}

@media only screen and (max-width: 1200px) {
    .movie {
        flex: 1 0 33%
    }
}

@media only screen and (max-width: 900px) {
    .movie {
        flex: 1 0 50%
    }
}

.movie__info {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    width: 282px;
    height: 100%;
    position: absolute;
    bottom: 0;
    left: 0;
    color: #fff;

    text-align: left;
    -webkit-transition: all 0.3s;
    -moz-transition: all 0.3s;
    -ms-transition: all 0.3s;
    -o-transition: all 0.3s;

}

.movie__title {
    padding: 10px;
    margin: 0;
    width: 100%;
    background: linear-gradient(rgba(0, 0, 0, 0), #000);
    transition: all 0.3s;
}

.movie__additional {
    margin: 0;
    padding: 0 10px;
    color: #fff;
    font-weight: 700;
    bottom: 100px;
    height: 0;
    overflow: hidden;
    transition: height 0.3s;
    background: rgba(0, 0, 0, 1);
}

@media only screen and (max-width: 600px) {
    .movie {
        flex: 1 0 100%
    }
}
