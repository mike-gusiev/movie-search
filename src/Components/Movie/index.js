import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom'

import './movie.css'

export default class Movie extends Component {

    constructor(props) {
        super(props);
        this.timer = null;
        this.ref = React.createRef();
        this.state = {
            shownItems: []
        }
    }

    componentDidMount() {
        window.addEventListener('scroll', this.handleScroll.bind(this));
    }

    componentDidUpdate() {
        this.debounceViewedMovies(1000);
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.handleScroll.bind(this));
    }

    checkViewedMovies = (element) => {
        if (!element) return;

        const rect = element.getBoundingClientRect();

        if (rect.top > 0 && rect.top < window.innerHeight) {
            if (!this.state.shownItems.includes(element.textContent)) {
                this.setState({
                    shownItems: [...this.state.shownItems, element.textContent]
                });
                console.log(`Movie view: ${element.textContent}`);
            }
        }
    }

    debounceViewedMovies(timeout) {
        clearTimeout(this.timer);
        this.timer = setTimeout(() => this.checkViewedMovies(this.ref.current), timeout);
    }

    handleScroll() {
        this.debounceViewedMovies(100);
    }

    render() {
        const {poster, title, year, id} = this.props;
        return (
            <div className="movie" ref={this.ref}>
                <Link to={{pathname: `/movie/${id}`, state: {poster, title, year}}}>
                    <div className="movie__container">
                        {poster === 'N/A' ? (
                            <img src="http://www.citypages.com/img/movie-placeholder.gif" alt="placeholder"/>
                        ) : (
                            <img src={poster} alt={title}/>
                        )}
                        <div className="movie__info">
                            <h2 className="movie__title">{title}</h2>
                            <p className="movie__additional">{year}</p>
                        </div>
                    </div>
                </Link>
            </div>
        )
    }
}

Movie.propTypes = {
    poster: PropTypes.string,
    title: PropTypes.string,
    year: PropTypes.string,
    id: PropTypes.string
};
