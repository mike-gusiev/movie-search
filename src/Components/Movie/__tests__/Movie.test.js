import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

import Movie from '../index';

const defaultProps = {
    poster: 'http://www.citypages.com/img/movie-placeholder.gif',
    title: 'image placeholder'
};

const setup = (props = defaultProps) => {
    const wrapper = shallow(<Movie {...props} />);
    return {
        wrapper,
        props
    };
};

describe('TEST Movie component ', () => {

    test('TEST Should match the snapshot.', () => {
        const {wrapper} = setup();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

});
