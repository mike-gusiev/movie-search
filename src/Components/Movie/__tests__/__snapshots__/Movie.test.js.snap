// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST Movie component  TEST Should match the snapshot. 1`] = `
<div
  className="movie"
>
  <Link
    replace={false}
    to={
      Object {
        "pathname": "/movie/undefined",
        "state": Object {
          "poster": "http://www.citypages.com/img/movie-placeholder.gif",
          "title": "image placeholder",
          "year": undefined,
        },
      }
    }
  >
    <div
      className="movie__container"
    >
      <img
        alt="image placeholder"
        src="http://www.citypages.com/img/movie-placeholder.gif"
      />
      <div
        className="movie__info"
      >
        <h2
          className="movie__title"
        >
          image placeholder
        </h2>
        <p
          className="movie__additional"
        />
      </div>
    </div>
  </Link>
</div>
`;
